File: Linked_Lists/206_reverse_linked_list/206.cpp
#include <iostream>
#include <vector>
#include "../list.cpp"

// LIST NODE DEFINITION
// struct SingleListNode {
//     int val;
//     SingleListNode *next;
//     SingleListNode() : val(0), next(nullptr) {}
//     SingleListNode(int x) : val(x), next(nullptr) {}
//     SingleListNode(int x, SingleListNode *next) : val(x), next(next) {}
// };

SingleListNode *reverseList(SingleListNode *head) {
    
    if (head == nullptr) {
        return nullptr;
    } else if (head -> next == nullptr) {
        return head;
    }

    SingleListNode *lastNode = reverseList(head -> next);
    head -> next -> next = head;
    head -> next = nullptr;
    return lastNode;
}


int main() {

    std::vector<int> array = {1, 2, 3, 4, 5, 6};

    SingleListNode *root = buildTree(array);

    std::cout << "list : ";
    stdout_list(root);

    root = reverseList(root);
    std::cout << "result : ";
    stdout_list(root);

    return 0;
}

File: Linked_Lists/206_reverse_linked_list/206.rs
mod list;


// List Definition
// #[derive(Debug)]
// struct SingleListNode {
//     val: i32,
//     next: Option<Box<SingleListNode>>,
// }

// impl SingleListNode {
//     fn new(val: i32) -> SingleListNode {
//         return SingleListNode{ val: val, next: None};
//     }
// }



fn reverse_list(head: Option<Box<list::SingleListNode>>) -> Option<Box<list::SingleListNode>> {
    
    let mut prev_node = None;
    let mut next_node = head;

    while let Some(mut node) = next_node {
        next_node = node.next;
        node.next = prev_node;
        prev_node = Some(node);
    }
    return prev_node;
}



fn main() {

    let array: Vec<i32> = vec!(1, 2, 3, 4, 5, 6, 7, 8);
    let head = list::build_list(array);

    list::print_nodes(&head);

    let head = reverse_list(head);

    list::print_nodes(&head);
}